### ISSUE TITLE
Optimization of Eigen backend

### ISSUE THREAD

**Issue by kaorahi** [lightvector/KataGo#288]
ref. https://github.com/lightvector/KataGo/issues/14#issuecomment-668226136

As a first step, v1.5.0 is 13 times slower than this "naive loop" version. So something is quite inefficient. I guess delayed evaluation of pad() may be harmful.

https://github.com/lightvector/KataGo/compare/1.5.0...kaorahi:eigen_naive_loop

**Comment by kaorahi** [comment]
This "handmade padding" version is 19 times faster than v1.5.0. What is the right way of padding with Eigen??

https://github.com/lightvector/KataGo/compare/1.5.0...kaorahi:eigen_handmade_padding

**Comment by lightvector** [comment]
That's an impressively large difference...

**Comment by sanderland** [comment]
There are some very interesting links in here: https://github.com/Dobiasd/frugally-deep/issues/167
It seems there's an optimized Eigen-based convolution in tensorflow, as well as the option to chain a few function calls to flatten the kernel and matrix-multiply the whole thing.

**Comment by kaorahi** [comment]
Thx! Yes. I'm just debugging this. x145 speed up from 1.5.0, though the output is wrong :p

~~~
  void apply(CONSTTENSORMAP4* input, TENSORMAP4* output, bool accumulate) const {
    assert(output->dimension(0) == outChannels);
    int i_c = input->dimension(0), i_w = input->dimension(1);
    int i_h = input->dimension(2), i_n = input->dimension(3);
    int k_w = kernel.dimension(0), k_h = kernel.dimension(1);
    int k_ic = kernel.dimension(2), k_oc = kernel.dimension(3);
    assert(i_c == k_ic);
    int vector_len = k_ic * k_w * k_h;
    Eigen::array<int, 4> matched_order({2, 0, 1, 3});
    Eigen::array<int, 2> as_row_vectors({k_oc, vector_len});
    Eigen::array<int, 2> as_col_vectors({vector_len, i_w * i_h * i_n});
    Eigen::array<Eigen::IndexPair<int>, 1> as_inner_products
      = {Eigen::IndexPair<int> (1, 0)};
    Eigen::array<int, 4> output_shape({k_oc, i_w, i_h, i_n});
    auto cooked_kernel = kernel.shuffle(matched_order)
      .reshape(as_row_vectors);
    auto cooked_input = input->extract_image_patches(k_w, k_h)
      .reshape(as_col_vectors);
    auto convolution = cooked_kernel
      .contract(cooked_input, as_inner_products)
      .reshape(output_shape);
    if(accumulate)
      *output += convolution;
    else
      *output = convolution;
  }
~~~

**Comment by sanderland** [comment]
Yes, the `extract_image_patches` is what I found as well, I think here - https://stackoverflow.com/questions/44425862/tensorflow-convolution-code-optimization

**Comment by kaorahi** [comment]
"image_patches" version: x170 speed up from 1.5.0

https://github.com/lightvector/KataGo/compare/1.5.0...kaorahi:eigen_image_patches

(gprof for `katago benchmark -model net/kata_96x6.txt.gz -config gtp_my.cfg -t 6 -n 1 -v 10`)
~~~
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls   s/call   s/call  name    
 37.58      0.56     0.56     1084     0.00     0.00  Eigen::internal::gebp_kernel<float, float, long, Eigen::internal::blas_data_mapper<float, long, 0, 0>, 8, 4, false, false>::operator()(Eigen::internal::blas_data_mapper<float, long, 0, 0> const&, float const*, float const*, long, long, long, float, long, long, long, long)
 31.54      1.03     0.47 268501739     0.00     0.00  Eigen::internal::TensorIntDivisor<long, false>::divide(long) const
 14.77      1.25     0.22 22375853     0.00     0.00  float __vector(4) Eigen::TensorEvaluator<Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const> const, Eigen::DefaultDevice>::packet<0>(long) const
  3.36      1.30     0.05      216     0.00     0.00  BatchNormLayer::apply(bool, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const*, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer>*, Eigen::TensorMap<Eigen::Tensor<float, 3, 0, long>, 0, Eigen::MakePointer> const*) const
  2.68      1.34     0.04     1063     0.00     0.00  Eigen::internal::gemm_pack_rhs<float, long, Eigen::internal::TensorContractionSubMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<std::array<int, 2ul> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const> const> const, Eigen::DefaultDevice>, std::array<long, 1ul>, std::array<long, 1ul>, 4, true, false, 0>, 4, 0, false, false>::operator()(float*, Eigen::internal::TensorContractionSubMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<std::array<int, 2ul> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const> const> const, Eigen::DefaultDevice>, std::array<long, 1ul>, std::array<long, 1ul>, 4, true, false, 0> const&, long, long, long, long)
~~~

It is still less than 9% of the speed of another CPU-only implementation, [libtensorflow backend](https://github.com/lightvector/KataGo/issues/14#issuecomment-549142392). So we have a large room of further optimizations.

(libtensorflow backend is almost a copy of @y-ich's work (#67). I understand nothing actually.)

**Comment by sanderland** [comment]
Nice work @kaorahi, getting 3 visits/second on 15b with this image_patches version, rather than 0.01 or whatever before.

**Comment by kaorahi** [comment]
"multi_thread" version: x380 speed up from 1.5.0 (19% of libtensorflow backend). 35 visits/s in benchmark and 20-30 visits/s in GUI with 6b net in my case.

https://github.com/kaorahi/KataGo/compare/eigen_image_patches...eigen_multi_thread

`num_eigen_threads` is hard coded at present.

(gprof for `katago benchmark -model net/kata_96x6.txt.gz -config gtp_my.cfg -t 6 -n 1 -v 10`)
~~~
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  ms/call  ms/call  name    
 30.99      0.38     0.38 26311371     0.00     0.00  Eigen::internal::TensorIntDivisor<long, false>::divide(long) const
 23.14      0.66     0.28     8141     0.03     0.03  Eigen::internal::gebp_kernel<float, float, long, Eigen::internal::blas_data_mapper<float, long, 0, 0>, 8, 4, false, false>::operator()(Eigen::internal::blas_data_mapper<float, long, 0, 0> const&, float const*, float const*, long, long, long, float, long, long, long, long)
 15.70      0.85     0.19      216     0.88     0.88  BatchNormLayer::apply(bool, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const*, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer>*, Eigen::TensorMap<Eigen::Tensor<float, 3, 0, long>, 0, Eigen::MakePointer> const*) const
 14.05      1.02     0.17  2238239     0.00     0.00  float __vector(4) Eigen::TensorEvaluator<Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const> const, Eigen::ThreadPoolDevice>::packet<0>(long) const
  3.31      1.06     0.04        1    40.00    40.00  ScoreValue::initTables()
~~~

**Comment by kaorahi** [comment]
"borrow_tensorflow" version: x1400 speed up from 1.5.0 (70% of libtensorflow backend). Now 15b net is usable for me. I get 19 visits/s in benchmark and 10 visits/s in GUI with 15b net.

I noticed tensorflow has a sophisticated convolver based on "virtual tensor". So I borrowed it shamelessly as its reimplementaion seems too hard. @lightvector, is this acceptable? (I also note that the previous "multi_thread" version attains x730 speed up from 1.5.0 if I give the compiler option `-DCMAKE_CXX_FLAGS='-march=native -mfpmath=both'`.)

https://github.com/lightvector/KataGo/compare/1.5.0...kaorahi:eigen_borrow_tensorflow

You need to copy two source files from tensorflow. See `src/cpp/neuralnet/from_tensorflow/README` for the instructions. You may also want to modify `num_eigen_threads = 8` in `src/cpp/neuralnet/eigenbackend.cpp`.

(gprof for `katago benchmark -model net/kata_96x6.txt.gz -config gtp_my.cfg -t 6 -n 1 -v 10`)
~~~
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  ms/call  ms/call  name    
 32.50      0.13     0.13     8067     0.02     0.02  Eigen::internal::gebp_kernel<float, float, long, Eigen::internal::blas_data_mapper<float, long, 0, 0, 1>, 24, 4, false, false>::operator()(Eigen::internal::blas_data_mapper<float, long, 0, 0, 1> const&, float const*, float const*, long, long, long, float, long, long, long, long)
 27.50      0.24     0.11       87     1.26     1.26  BatchNormLayer::apply(bool, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const*, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer>*, Eigen::TensorMap<Eigen::Tensor<float, 3, 0, long>, 0, Eigen::MakePointer> const*) const
  7.50      0.27     0.03     2000     0.02     0.02  Eigen::internal::gemm_pack_rhs<float, long, Eigen::internal::TensorContractionSubMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const> const> const, Eigen::ThreadPoolDevice>, std::array<long, 1ul>, std::array<long, 1ul>, 8, true, false, 0, Eigen::MakePointer>, 4, 0, false, false>::operator()(float*, Eigen::internal::TensorContractionSubMapper<float, long, 0, Eigen::TensorEvaluator<Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float, 4, 0, long>, 0, Eigen::MakePointer> const> const> const, Eigen::ThreadPoolDevice>, std::array<long, 1ul>, std::array<long, 1ul>, 8, true, false, 0, Eigen::MakePointer> const&, long, long, long, long) const
  7.50      0.30     0.03        1    30.00    30.00  ScoreValue::initTables()
  5.00      0.32     0.02  1030599     0.00     0.00  checkWeightFinite(float, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
~~~

**Comment by lightvector** [comment]
I haven't looked at it in detail, but: it's only acceptable if you've investigated the license and determined it to be compatible. Is that the case?

Also if you want to clean all this up and submit it as a pull request, please make sure to clean up the code, the placement of files, formatting of commit messages, etc and make it match the existing style of the code. External code should go into cpp/external along with documentation about that code's license and the cmake configuration should be set up so that besides specifying EIGEN for the backend, the user does not need to perform any additional steps.

Thanks for working on such major improvements!

**Comment by lightvector** [comment]
Alternatively, once you get to something you're happy with and you're done optimizing here, you can also just submit what you have, and I'll clean it up as I merge it in - I'm happy to do that as well!

**Comment by lightvector** [comment]
@kaorahi 
Please take a look at the following branch, where I did try porting the OpenCL winograd implementation to the Eigen implementation.
https://github.com/lightvector/KataGo/tree/eigen-winograd

I forked your code from the point of image patches - I have not attempted to merge any of your later optimizations such as adding the thread pool or other stuff - but hopefully these later optimizations can also help the winograd implementation too?

For me, the winograd implementation, even though a lot of it is just naive for loops that ignore memory locality (which I am not sure how to do better on), it achieves a factor of 3 improvement over the image patches. Do you think it would be possible to see if that factor improvement can be preserved as one adds on the later improvements you implemented as well? Also, do you have any ideas on how one might optimize the transform/untransform steps themselves?

**Comment by lightvector** [comment]
What would be very cool of course would be if any of the libtensorflow optimization ideas are also applicable to the winograd implementation. Maybe not, but if they are, one could hope that enough of the factor 3 can be preserved to then to beat the libtensorflow version as well.

**Comment by kaorahi** [comment]
It's doubly faster! The factor is 1.94 in my case with `-march=native -mfpmath=both` for both versions.

`clock()` says the transform/untransform steps took roughly comparable time with the core elementwise multiplications. I'm afraid multi-threading may not work for the former.

It seems that the key idea of TensorFlow's SpatialConvolution is "virtual tensor" that avoids allocation of large tensors returned by Eigen's extract_image_patches.

**Comment by lightvector** [comment]
A thought: what if rather than using Eigen's thread pool, we tune `numNNServerThreadsPerModel`? This would "multithread" the transform and untransform steps too. Can comparable performance be reached? Also, perhaps there is a way to rearrange transform and untransform loops to be more efficient, or to vectorize them, or something.

**Comment by lightvector** [comment]
Okay, I rearranged winograd loop order for better memory locality and got a mild improvement. Then, parallelizing the backend simply by making more server threads seems to make it competitive with the thread pooling and/or tensorflow approaches, at least for me, (and the tensorflow thing is super-annoying to make compile). So I'm going with winograd for now.

I cleaned up everything and pushed it all to master. 

`numNNServerThreadsPerModel`, if left blank, will now default in the Eigen backend to be equal to numSearchThreads. The benchmark is also aware of this when tuning between different numbers of threads. Feel free to try it out!

**Comment by lightvector** [comment]
As for `-march=native`, this won't work for a precompiled release, but it seems like at part of the gain is captured via just AVX2. So I could release both a vanilla version and a `-mavx2` version which should work on most Intel CPUs in the last several years.

**Comment by kaorahi** [comment]
It runs nearly as fast as Leela Zero with Eigen backend (faster than my tensorflow-based versions). Thanks for such rapid developments. I'm happy to get practical CPU-only implementation officially.

**Comment by lp200** [comment]
If using -mavx2, it is recommended to add -mfma
